base/c.jl

```diff
@@ -503,3 +503,183 @@ end
 macro ccallable(rt, def)
     expand_ccallable(rt, def)
 end
+
+module CcallMacro
+
+struct NoType end
+struct CcallError{T <: AbstractString} <: Exception
+    msg::T
+end
```

Comment:
My thinking was that using a more specific error makes it easier to debug. Using the same few exception types everywhere, increasing the likelihood of catching something other than what you're fishing for. If my code is broken and tries to preform an action that raises an `ArgumentError`, I don't want to accidentally catch _that_ when testing my errors.

In general, I think having specific exception types for specific scenarios is always better. However, any error that actually happens in macro expansion will get turned into a `LoadError` anyway. You only ever get `CcallError` if you run the internal code generation functions directly.

Reply:
I still think this is not an improvement, but you're the maintainers, so "fixed" in https://github.com/JuliaLang/julia/pull/32748/commits/bbd6c5af06e54f25126de65d9e0cae7c0d4cd846

Reply:
Thanks. I do understand where you're coming from here, but there's always a tradeoff between user-visible complexity (increased by adding new types) vs the benefit of being able to match a specific error. In this case, I can't really imagine how matching `CcallError` specifically would be helpful. Especially given that it occurs only at macro expansion time.

Also keep in mind that it's still an open question what will become of error handling in Julia in the longer term. We don't even have syntax for matching exception types yet. (Maybe we could do more powerful and expressive matching on exception structure rather than type, in the style of Elixir or some such thing? Or have something completely different? The design space is still wide open there.) In the meantime, `Base` is relatively conservative about adding a lot of exception types.